### PR TITLE
fix: convert recurring-swap failure watcher to startListening pattern and gate push subscriptions on kill-switch

### PR DESCRIPTION
**Pull request by ruijia1in**
iOS: 9119
Android: 9120

## Description

**Ticket: [CP-]**

Please provide:

Wire startRecurringFailureWatcher via a startListening(onRehydrationComplete) effect instead of a bare call, matching addAppListeners' bootstrap shape, and unsubscribe after first fire so the cache subscriber wires exactly once.

Gate the per-order push-notification subscription pass on the SWAP_RECURRING kill-switch (selectIsRecurringSwapsBlocked). The gate is read LIVE on each matching cache event via a reader (`() => selectIsRecurringSwapsBlocked(listenerApi.getState())`) rather than snapshotted at wire time — the watcher wires on onRehydrationComplete, before the first PostHog fetch lands, so a snapshot would freeze the previously-persisted value (blocked on a fresh install) for the whole session even after the feature is enabled. The selector read is deferred to the use site (after handleQueryCacheEvent's early-returns), so it is only paid on RECURRING_SCHEDULES success events, not on every cache update. The failure snackbar and pending-action reconciliation still run regardless.

## Screenshots/Videos

Include relevant screenshots or screen recordings of iOS and Android.

## Testing

**Dev Testing (if applicable)**

- Provide steps to test the happy path of your feature
- Provide steps to test edge cases and error states
- Trigger a build on bitrise and reference it here
- Move the ticket into the "Testing" column on Jira

**QA Testing (if applicable)**

- Provide instructions for QA to test this feature thoroughly
- State expected behavior / acceptance criteria

## Checklist

Please check all that apply (if applicable)

- [ ] I have performed a self-review of my code
- [ ] I have verified the code works
- [ ] I have included screenshots / videos of android and ios
- [ ] I have added testing steps
- [ ] I have added/updated necessary unit tests
- [ ] I have updated the documentation

### CODE REVIEW

File: packages/core-mobile/app/new/features/recurringSwap/store/listeners.ts [hunk 1]
```diff
@@ -444,5 +464,22 @@ export function addRecurringSwapListeners(
...
+  // the first fire so the cache subscriber is attached exactly once even if
+  // `onRehydrationComplete` were ever re-dispatched.
+  startListening({
+    actionCreator: onRehydrationComplete,
+    effect: (_, listenerApi) => {
+      listenerApi.unsubscribe()
+      // Snapshot the kill-switch at wire time and gate the push-notification
+      // subscription pass on it — when recurring swaps are blocked there's no
+      // toggle/banner/management screen, so nothing to subscribe the device for.
+      const isRecurringSubscriptionBlocked = selectIsRecurringSwapsBlocked(
```
Comment (atn4z7) [inline packages/core-mobile/app/new/features/recurringSwap/store/listeners.ts]: This snapshots the flag once at wire time, but every other gate for recurring swaps reads it live via `useSelector` (the swap toggle, the Activity banner, the management-screen route guard). Because this effect runs on `onRehydrationComplete`, before the first live PostHog fetch lands, the value it freezes is whatever was persisted from the previous session, and it never re-reads for the rest of the session.

Two cases where that bites:
- Fresh install: no persisted flag yet, so this reads the default (`false` = blocked). Even after the 30s periodic fetch turns it on, the UI exposes the feature, and the user creates a schedule, the push-subscribe pass stays gated off until the next app restart.
- Kill-switch flipped off mid-session: the UI hides immediately but this keeps subscribing new orders on the stale value.

Could we pass a reader instead of a captured boolean so it reads live like the UI? e.g. `startRecurringFailureWatcher(() => selectIsRecurringSwapsBlocked(listenerApi.getState()))` and evaluate it inside the subscribe callback. `listenerApi.getState` returns current state on each call, so each cache event re-reads the flag.

One caveat either way: this only gates new subscribes. A device already subscribed while the flag was on keeps getting pushes until the backend stops sending, since there's no client-side unsubscribe. So this is a correctness improvement, but the full kill still relies on the backend honoring the flag too.

File: packages/core-mobile/app/new/features/recurringSwap/store/listeners.test.ts [hunk 2]
```diff
@@ -641,13 +654,28 @@ describe('addRecurringSwapListeners', () => {
     })
   })
 
-  it('starts the React Query cache subscriber', () => {
+  it('starts the React Query cache subscriber on rehydration', () => {
```
Comment (atn4z7) [inline packages/core-mobile/app/new/features/recurringSwap/store/listeners.test.ts:770]: This covers that the subscriber wires on rehydration, but nothing asserts the gate this PR adds. `subscribeForRecurringSwap` and `registerAndGetDeviceArn` aren't mocked, and there's no case for `blocked = true` skipping the per-order subscribe pass (or `blocked = false` running it). Could we add both branches so the gating behavior is locked in?

**Review by atn4z7 — COMMENTED** [review]
Nice PR.

File: packages/core-mobile/app/new/features/recurringSwap/store/listeners.ts [hunk 3]
```diff
@@ -387,15 +391,20 @@ function handleQueryCacheEvent(event: QueryCacheNotifyEvent): void {
...
+    // the device subscribed for. Kept out of the synchronous cache-subscriber
+    // call path; failures inside the function are logged but never bubble up —
+    // the next refetch retries any order that didn't make it into the persisted
+    // subscribed-set.
+    if (!isRecurringSubscriptionBlocked) {
+      ensureOrderSubscriptions(ownerAddress, chainId, schedules ?? []).catch(
+        err => {
+          Logger.error('[RecurringSwap] ensureOrderSubscriptions threw', err)
+        }
+      )
```
[the remark this answers is not in the document]
Comment (ruijia1in) [reply packages/core-mobile/app/new/features/recurringSwap/store/listeners.ts]: Done in 54614fffb. Added `void` on the fire-and-forget, with an inline `// eslint-disable-next-line no-void` — this repo's base config warns on `no-void`, so I matched the existing precedent in `MarketChart.tsx` (`void chartHtmlAsset.downloadAsync().catch(...)`) rather than introduce a new warning.

File: packages/core-mobile/app/new/features/recurringSwap/store/listeners.test.ts [hunk 4]
```diff
@@ -107,12 +107,24 @@ jest.mock('contexts/ReactQueryProvider', () => ({
   }
 }))
 
+// ─── posthog selector mock ────────────────────────────────────────────────────
+// The SWAP_RECURRING kill-switch drives whether the failure watcher wires up
+// the push-notification subscription pass. `listeners.ts` only consumes
+// `selectIsRecurringSwapsBlocked` from this barrel, so a minimal stub avoids
+// pulling the real module's heavy import graph (WalletFactory, etc.) into the
```
[the remark this answers is not in the document]
Comment (ruijia1in) [reply packages/core-mobile/app/new/features/recurringSwap/store/listeners.test.ts:114]: Added in 54614fffb. `subscribeForRecurringSwap` and `registerAndGetDeviceArn` are now mocked, with tests asserting: blocked=true → neither is called; blocked=false → subscribes with `{ orderId, deviceArn }`; and blocked=true still fires the failure snackbar (the gate only affects the subscribe pass).

File: packages/core-mobile/app/new/features/recurringSwap/store/listeners.ts (hunk 1 above)
Comment (ruijia1in) [reply packages/core-mobile/app/new/features/recurringSwap/store/listeners.ts]: Good catch — fixed in 54614fffb. Switched from the wire-time snapshot to a live reader: `startRecurringFailureWatcher` now takes `getIsRecurringSubscriptionBlocked: () => boolean` and evaluates it on every cache event, wired as `() => selectIsRecurringSwapsBlocked(listenerApi.getState())` exactly as you suggested. That fixes both the fresh-install freeze (watcher wires before the first PostHog fetch) and the mid-session flip.

Agree on your caveat too — this only gates *new* subscribes; a device already subscribed keeps receiving pushes until the backend stops sending, since there's no client-side unsubscribe. So the full kill still depends on the backend honoring the flag. Left that behavior as-is since it's out of scope for this PR.

File: packages/core-mobile/app/new/features/recurringSwap/store/listeners.test.ts (hunk 2 above)
Comment (ruijia1in) [reply packages/core-mobile/app/new/features/recurringSwap/store/listeners.test.ts:770]: Added in 54614fffb. Both branches are now covered, plus a live re-read case: a mid-session flip to blocked gates the *next* refetch (this would fail against the old snapshot). `subscribeForRecurringSwap` and `registerAndGetDeviceArn` are mocked, and there are two additional tests that go through the real `addRecurringSwapListeners` → `onRehydrationComplete` wiring and assert `getState()` is consulted per event on both branches.

**Review by atn4z7 — DISMISSED** [review]

File: packages/core-mobile/app/new/features/recurringSwap/store/listeners.ts [hunk 5]
```diff
@@ -408,9 +423,27 @@ function handleQueryCacheEvent(event: QueryCacheNotifyEvent): void {
...
-export function startRecurringFailureWatcher(): () => void {
-  return queryClient.getQueryCache().subscribe(handleQueryCacheEvent)
+export function startRecurringFailureWatcher(
+  getIsRecurringSubscriptionBlocked: () => boolean
+): () => void {
+  return queryClient
+    .getQueryCache()
+    .subscribe(event =>
+      handleQueryCacheEvent(event, getIsRecurringSubscriptionBlocked())
+    )
```
[the remark this answers is not in the document]
Comment (ruijia1in) [reply packages/core-mobile/app/new/features/recurringSwap/store/listeners.ts:459]: Fixed in 9bc73539a. The reader is now passed into `handleQueryCacheEvent` un-invoked and called at the use site, after the early-returns (queryKey / type / status guards). So `getState()`/the selector is only read for RECURRING_SCHEDULES success events — unrelated high-frequency cache updates return before any Redux read.

File: packages/core-mobile/app/new/features/recurringSwap/store/listeners.ts [hunk 6]
```diff
@@ -408,9 +423,27 @@ function handleQueryCacheEvent(event: QueryCacheNotifyEvent): void {
...
+ * `getIsRecurringSubscriptionBlocked` is a reader invoked LIVE on every cache
+ * event — NOT a boolean snapshotted at wire time. It gates the per-order
+ * push-notification subscription pass on the SWAP_RECURRING kill-switch, and
+ * reading live matches how every other recurring-swap gate consumes the flag
+ * (the swap toggle, the Activity banner, the management-screen route guard all
+ * read it via `useSelector`). The watcher is wired on `onRehydrationComplete`,
+ * before the first live PostHog fetch lands — snapshotting there would freeze
+ * the previous session's persisted value (default `false`/blocked on a fresh
+ * install) and never re-read, so a schedule created after the flag flips on
+ * mid-session would stay unsubscribed until the next app restart. The failure
```
[the remark this answers is not in the document]
Comment (ruijia1in) [reply packages/core-mobile/app/new/features/recurringSwap/store/listeners.ts]: Clarified in 9bc73539a. Reworded the doc comment to distinguish the two: on a fresh install nothing is persisted, so the raw `SWAP_RECURRING` gate reads as off and `selectIsRecurringSwapsBlocked` returns `true` (blocked). The point is that a wire-time snapshot would freeze that blocked value for the whole session even after the periodic fetch enables the feature.

File: packages/core-mobile/app/new/features/recurringSwap/store/listeners.ts [hunk 7]
```diff
@@ -444,5 +477,24 @@ export function addRecurringSwapListeners(
...
+      listenerApi.unsubscribe()
+      // Pass a LIVE reader for the kill-switch, not a wire-time snapshot:
+      // this effect runs on rehydration, before the first PostHog fetch, so a
+      // captured boolean would freeze the previous session's persisted value
+      // and never re-read. `listenerApi.getState()` returns current state on
+      // each call, so the subscribe pass re-evaluates the flag on every cache
+      // event — matching how the UI gates read it live.
+      startRecurringFailureWatcher(() =>
+        selectIsRecurringSwapsBlocked(listenerApi.getState())
+      )
```
[the remark this answers is not in the document]
Comment (ruijia1in) [reply packages/core-mobile/app/new/features/recurringSwap/store/listeners.ts:510]: Updated the PR description in 9bc73539a to match the implementation: the gate is read LIVE per matching cache event (via a reader), not snapshotted. The description now explains why (watcher wires on onRehydrationComplete, before the first PostHog fetch) and that the selector read is deferred to the use site so it's only paid on RECURRING_SCHEDULES success events.

File: packages/core-mobile/app/new/features/recurringSwap/store/listeners.test.ts [hunk 8]
```diff
@@ -641,13 +767,71 @@ describe('addRecurringSwapListeners', () => {
...
+    call?.[0].effect(undefined, { unsubscribe, getState })
+
+    // Fires exactly once and drops itself so the subscriber can't double-wire.
+    expect(unsubscribe).toHaveBeenCalledTimes(1)
     expect(mockCacheSubscribers).toHaveLength(1)
   })
+
+  // atn4z7's review: the wired subscriber must read the kill-switch LIVE from
+  // `listenerApi.getState()` on every cache event, not snapshot it at wire
+  // time. These two cases lock in both branches through the real wiring path.
```
[the remark this answers is not in the document]
Comment (ruijia1in) [reply packages/core-mobile/app/new/features/recurringSwap/store/listeners.test.ts]: Done in 9bc73539a. Dropped the reviewer handle — the comment now just states the contract (the wired subscriber must read the kill-switch live from `listenerApi.getState()` per matching event, not snapshot it).

**Review by atn4z7 — APPROVED** [review]

**Review by B0Y3R-AVA — APPROVED** [review]
dev tested
